contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,807 @@
+
+        // slash
+        uint256 slashAmount = IStakeCredit(valInfo.creditContract).slash(downtimeSlashAmount);
+        uint256 jailUntil = block.timestamp + downtimeJailTime;
+        _jailValidator(valInfo, jailUntil);
+
+        emit ValidatorSlashed(operatorAddress, jailUntil, slashAmount, SlashType.DownTime);
+
+        _sync(valInfo.creditContract, operatorAddress);
+    }
+
+    function maliciousVoteSlash(bytes calldata _voteAddr) external onlySlash {
+        address operatorAddress = _voteToOperator[_voteAddr];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.creditContract != address(0), "VALIDATOR_NOT_EXIST");
+
+        // check if can be jailed
+        // there are two slots for each day
+        uint256 dayIndex1 = (block.timestamp / 1 days) * 2;
+        uint256 dayIndex2 = dayIndex1 + 1;
+        if (_canBeJailed(dayIndex1)) {
+            _setJailed(dayIndex1);
+        } else if (_canBeJailed(dayIndex2)) {
+            _setJailed(dayIndex2);
+        } else {
+            return;
+        }
+
+        // slash
+        (bool canSlash, uint256 jailUntil) = _checkSlashRecord(operatorAddress, SlashType.MaliciousVote);
+        require(canSlash, "ALREADY_SLASHED");
+        uint256 slashAmount = IStakeCredit(valInfo.creditContract).slash(doubleSignSlashAmount);
+        _jailValidator(valInfo, jailUntil);
+
+        emit ValidatorSlashed(operatorAddress, jailUntil, slashAmount, SlashType.MaliciousVote);
+
+        _sync(valInfo.creditContract, operatorAddress);
+    }
+
+    function doubleSignSlash(address consensusAddress) external onlySlash {
+        address operatorAddress = _consensusToOperator[consensusAddress];
+        require(operatorAddress != address(0), "INVALID_CONSENSUS_ADDRESS"); // should never happen
+        Validator storage valInfo = _validators[operatorAddress];
+        require(valInfo.creditContract != address(0), "VALIDATOR_NOT_EXIST");
+
+        // check if can be jailed
+        // there are two slots for each day
+        uint256 dayIndex1 = (block.timestamp / 1 days) * 2;
+        uint256 dayIndex2 = dayIndex1 + 1;
+        if (_canBeJailed(dayIndex1)) {
+            _setJailed(dayIndex1);
+        } else if (_canBeJailed(dayIndex2)) {
+            _setJailed(dayIndex2);
+        } else {
+            revert("NO_JAIL_SLOT");
+        }
+
+        // slash
+        (bool canSlash, uint256 jailUntil) = _checkSlashRecord(operatorAddress, SlashType.DoubleSign);
+        require(canSlash, "ALREADY_SLASHED");
+        uint256 slashAmount = IStakeCredit(valInfo.creditContract).slash(doubleSignSlashAmount);
+        _jailValidator(valInfo, jailUntil);
+
+        emit ValidatorSlashed(operatorAddress, jailUntil, slashAmount, SlashType.DoubleSign);
+
+        _sync(valInfo.creditContract, operatorAddress);
+    }
+
+    /*----------------- gov -----------------*/
```

Comment:
The comment here is not the right place.

Reply:
Removed